src/Pos/Communication/Relay.hs

```diff
@@ -122,104 +130,214 @@ handleReqL
+                    send $ ReqMsg imTag ne
+                    whenJustM recv $ expectRight $
+                        \dt@DataMsg{..} -> handleDataL proxy dt
+
+    expectLeft call (Left msg) = call msg
+    expectLeft _ (Right _)     = throw UnexpectedData
+
+    expectRight _ (Left _)       = throw UnexpectedInv
+    expectRight call (Right msg) = call msg
+
 
 relayStubListeners
     :: ( WithLogger m
        , Bi (InvMsg key tag)
        , Bi (ReqMsg key tag)
        , Bi (DataMsg contents)
-       , Message (InvMsg key tag)
+       , Message (InvOrData tag key contents)
        , Message (ReqMsg key tag)
-       , Message (DataMsg contents)
        )
     => RelayProxy key tag contents -> ([ListenerSpec m], OutSpecs)
 relayStubListeners p = mergeLs
-    [ stubListenerOneMsg $ invMsgProxy p
-    , stubListenerOneMsg $ reqMsgProxy p
-    , stubListenerOneMsg $ dataMsgProxy p
+    [ stubListenerConv $ invDataMsgProxy p
+    , stubListenerConv $ reqMsgProxy p
     ]
+
+invCatchType :: RelayProxy key tag contents -> InvMsg key tag -> ()
+invCatchType _ _ = ()
+
+reqCatchType :: RelayProxy key tag contents -> ReqMsg key tag -> ()
+reqCatchType _ _ = ()
+reqMsgProxy :: RelayProxy key tag contents
+            -> Proxy (InvOrData tag key contents, ReqMsg key tag)
+reqMsgProxy _ = Proxy
+
+dataCatchType :: RelayProxy key tag contents -> DataMsg ontents -> ()
+dataCatchType _ _ = ()
+
+invDataMsgProxy :: RelayProxy key tag contents
+                -> Proxy (ReqMsg key tag, InvOrData tag key contents)
+invDataMsgProxy _ = Proxy
+
+
+addToRelayQueue :: forall tag key contents ssc m .
+                ( Bi (InvOrData tag key contents)
+                , Bi (ReqMsg key tag)
+                , Message (InvOrData tag key contents)
+                , Message (ReqMsg key tag)
+                , Buildable tag, Buildable key
+                , WorkMode ssc m, Bi NOP, Message NOP)
+                => InvOrData tag key contents -> m ()
+addToRelayQueue inv = do
+    queue <- ncInvPropagationQueue <$> getNodeContext
+    isFull <- atomically $ isFullTBQueue queue
+    if isFull then
+        logWarning $ "Propagation queue is full, no propagation"
+    else
+        atomically $ writeTBQueue queue (SomeInvMsg inv)
+
+relayWorkers :: forall ssc m .
+             ( Mockable Throw m
+             , WorkMode ssc m
+             , Bi NOP, Message NOP)
+             => ([WorkerSpec m], OutSpecs)
+relayWorkers = first (:[]) $ worker allOutSpecs $
+  \sendActions -> handleAll handleWE $ do
+    queue <- ncInvPropagationQueue <$> getNodeContext
+    forever $ do
+        inv <- atomically $ readTBQueue queue
+        case inv of
+            SomeInvMsg (i@(Left (InvMsg{..}))) -> do
+                logDebug $
+                    sformat ("Propagation data with keys: "%listJson%" and tag: "%build) imKeys imTag
+                converseToNeighbors sendActions (convHandler i)
+            SomeInvMsg (Right _) -> throw InvalidPropagationElement
```

Comment:
No error throwing!! Just log. As it's written now, worker would finish, which shouldn't happen.